proposals/0000-net-storage-reads-improvements.md

```diff
@@ -0,0 +1,105 @@
+
+Unfortunately, this network protocol is suffering from some issues:
+
+- It is not possible for the querier to check whether a key exists in the storage of the chain except by querying the value of that key. The reply will thus include the value of the key, only for that value to be discarded by the querier that isn't interested by it. This is a waste of bandwidth.
+- It is not possible for the querier to know whether a value in the storage of the chain has been modified between two blocks except by querying this value for both blocks and comparing them. Only a few storage values get modified in a block, and thus most of the time the comparison will be equal. This leads to a waste of bandwidth as the values have to be transferred.
+- While it is possible to ask for multiple specific storage keys at the same time, it is not possible to ask for a list of keys that start with a certain prefix. Due to the way FRAME works, storage keys are grouped by "prefix", for example all account balances start with the same prefix. It is thus a common necessity for a light client to obtain the list of all keys (and possibly their values) that start with a specific prefix. This is currently not possible except by performing multiple queries serially that "walk down" the trie.
+
+Once Polkadot and Kusama will have transitioned to `state_version = 1`, which modifies the format of the trie entries, it will be possible to generate Merkle proofs that contain only the hashes of values in the storage. Thanks to this, it is already possible to prove the existence of a key without sending its entire value (only its hash), or to prove that a value has changed or not between two blocks (by sending just their hashes).
+Thus, the only reason why aforementioned issues exist is because the existing networking messages don't give the possibility for the querier to query this. This is what this proposal aims at fixing.
+
+## Detailed Solution design
+The protobuf schema of the networking protocol can be found here: https://github.com/paritytech/substrate/blob/5b6519a7ff4a2d3cc424d78bc4830688f3b184c0/client/network/light/src/schema/light.v1.proto
+
+The proposal is to modify this protocol in this way:
+
+```diff
+@@ -11,6 +11,7 @@ message Request {
+                RemoteReadRequest remote_read_request = 2;
+                RemoteReadChildRequest remote_read_child_request = 4;
+                // Note: ids 3 and 5 were used in the past. It would be preferable to not re-use them.
++               RemoteReadRequestV2 remote_read_request_v2 = 6;
+        }
+ }
+ 
+@@ -48,6 +49,21 @@ message RemoteReadRequest {
+        repeated bytes keys = 3;
+ }
+ 
++message RemoteReadRequestV2 {
++       required bytes block = 1;
++       optional ChildTrieInfo child_trie_info = 2;  // Read from the main trie if missing.
++       repeated Key keys = 3;
++}
++
++message ChildTrieInfo {
++       enum ChildTrieNamespace {
++               DEFAULT = 1;
++       }
++
++       required bytes hash = 1;
```

Comment:
This could also be optional: if undefined, then the proof should contain the resolution of this child trie root in the top trie.
I feel like it is good to have it mandatory, as it makes it clear the reply should only contains a proof local to the child trie.
But at the same time it means that child trie proof will always be done in two request: query the child trie root (and possibly other content), then query the child trie.

Reply:
I don't understand your suggestion.
The request would provide a key and say that it is in a child trie, but without saying which one?

Reply:
same misunderstanding as in next comment. Cannot be optional since we don't have the child trie root.